### PR TITLE
[docs] Add syntax documentation for logical replication

### PR DESCRIPTION
**Pull request by dr0pdb**

### CODE REVIEW

File: docs/content/preview/api/ysql/the-sql-language/statements/_index.md [hunk 1]
```diff
@@ -45,6 +46,7 @@ The YSQL statements are compatible with the SQL dialect that PostgreSQL supports
 | [`CREATE OPERATOR`](ddl_create_operator) | Create an operator |
 | [`CREATE OPERATOR CLASS`](ddl_create_operator_class) | Create an operator class |
 | [`CREATE PROCEDURE`](ddl_create_procedure) | Create a procedure |
+| [`CREATE PUBLICATION`](ddl_create_publication) | Create a publication |
```
Comment (asrinivasanyb) [inline docs/content/preview/api/ysql/the-sql-language/statements/_index.md:49]: CREATE REPLICATION_SLOT is missing

File: docs/content/preview/api/ysql/the-sql-language/statements/streaming_create_repl_slot.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+### *slot_name*
+
+The name of the replication slot. The name of the replication slot must be **unique across all databases**.
+
+### *output_plugin*
+
+The name of the output plugin used for logical decoding.
+
+### NOEXPORT_SNAPSHOT / USE_SNAPSHOT
```
Comment (asrinivasanyb) [inline docs/content/preview/api/ysql/the-sql-language/statements/streaming_create_repl_slot.md:33]: We had discussed about not documenting the USE_SNAPSHOT option. We just document the NOEXPORT_SNAPSHOT

File: docs/content/preview/api/ysql/the-sql-language/statements/streaming_create_repl_slot.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+Decides what to do with the snapshot created during logical slot initialization.
+
+`USE_SNAPSHOT` will use the snapshot for the current transaction executing the command. This option must be used in a transaction, and `CREATE_REPLICATION_SLOT` must be the first command run in that transaction.
```
Comment (asrinivasanyb) [inline docs/content/preview/api/ysql/the-sql-language/statements/streaming_create_repl_slot.md:37]: The semantics of USE_SNAPSHOT in our case is that an initial snapshot will be supported for this stream - "will use the snapshot for the current transaction executing the command" does not hold for YB

Comment (asrinivasanyb) [reply docs/content/preview/api/ysql/the-sql-language/statements/streaming_create_repl_slot.md:37]: Doc from yb-admin https://docs.yugabyte.com/preview/admin/yb-admin/#change-data-capture-cdc-commands

**Review by asrinivasanyb — CHANGES_REQUESTED** [review]
USE_SNAPSHOT documentation needs correction

File: docs/content/preview/api/ysql/the-sql-language/statements/streaming_create_repl_slot.md (hunk 3 above)
Comment (dr0pdb) [reply docs/content/preview/api/ysql/the-sql-language/statements/streaming_create_repl_slot.md]: Done, thank you

File: docs/content/preview/api/ysql/the-sql-language/statements/ddl_alter_publication.md [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+Change the owner of the publication to the *new_owner*.
+
+### Permissions
+
+To alter a publication, the invoking user must own the publication. Adding a table to a publication additionally requires owning that table. To alter the owner, you must also be a direct or indirect member of the new owning role. The new owner must have `CREATE` privilege on the database. Also, the new owner of a `FOR ALL TABLES` publication must be a superuser. However, a superuser can change the ownership of a publication regardless of these restrictions.
+
+## Examples
+
+Assume that we already have 3 tables created within the database: `users`, `departments` and `employees`. Also, a publication `my_publication` has been created using the following command:
```
Comment (ddhodge) [inline docs/content/preview/api/ysql/the-sql-language/statements/ddl_alter_publication.md:56]:
```suggestion
Assume that you have created the tables `users`, `departments`, and `employees`.

To create a publication `my_publication`, use the following command:
```

File: docs/content/preview/api/ysql/the-sql-language/statements/ddl_alter_publication.md [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+```sql
+yugabyte=# CREATE PUBLICATION my_publication FOR TABLE employees;
+```
+
+Add tables `users` and `departments` to the publication `my_publication`.
```
Comment (ddhodge) [inline docs/content/preview/api/ysql/the-sql-language/statements/ddl_alter_publication.md:62]:
```suggestion
Add the `users` and `departments` tables to the publication `my_publication`:
```

File: docs/content/preview/api/ysql/the-sql-language/statements/ddl_alter_publication.md [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+```sql
+yugabyte=# ALTER PUBLICATION my_publication ADD TABLE users, departments;
+```
+
+Drop tables `employees`from the publication `my_publication`.
```
Comment (ddhodge) [inline docs/content/preview/api/ysql/the-sql-language/statements/ddl_alter_publication.md:68]:
```suggestion
Drop the `employees` table from the publication `my_publication`:
```

File: docs/content/preview/api/ysql/the-sql-language/statements/ddl_alter_publication.md [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+```sql
+yugabyte=# ALTER PUBLICATION my_publication DROP TABLE employees;
+```
+
+Set the tables to only `departments` in the publication `my_publication`. The table `users` will be removed from the publication as a result of the command.
```
Comment (ddhodge) [inline docs/content/preview/api/ysql/the-sql-language/statements/ddl_alter_publication.md:74]:
```suggestion
Set the `departments` table to be the only table in the publication `my_publication`. The table `users` will be removed from the publication as a result of the command.
```

File: docs/content/preview/api/ysql/the-sql-language/statements/ddl_create_publication.md [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+### Permissions
+
+To create a publication, the invoking user must have the `CREATE` privilege for the current database. (Superusers bypass this check.)
+
+To add a table to a publication, the invoking user must have ownership rights on the table. The `FOR ALL TABLES` clause requires the invoking user to be a **superuser**.
+
+### Limitations
+
+Currently to publish a subset of operations (create, update, delete, truncate) via a Publication is not supported.
```
Comment (ddhodge) [inline docs/content/preview/api/ysql/the-sql-language/statements/ddl_create_publication.md:42]:
```suggestion
Publishing a subset of operations (create, update, delete, truncate) via a Publication is not currently supported.
```

File: docs/content/preview/api/ysql/the-sql-language/statements/ddl_create_publication.md [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+## Examples
+
+### Publication with two tables
+
+```sql
+yugabyte=# CREATE PUBLICATION mypublication FOR TABLE users, departments;
+```
+
+In this example, the publication `mypublication` will consist of two tables, `users` and `departments`.
```
Comment (ddhodge) [inline docs/content/preview/api/ysql/the-sql-language/statements/ddl_create_publication.md:52]:
```suggestion
In this example, the publication `mypublication` consists of two tables, `users` and `departments`.
```

File: docs/content/preview/api/ysql/the-sql-language/statements/ddl_create_publication.md [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+(1 row)
+```
+
+### Publication for all tables
+
+```sql
+yugabyte=# CREATE PUBLICATION puballtables FOR ALL TABLES;
+```
+
+In this example, the publication `puballtables` will include all the current as well as future tables.
```
Comment (ddhodge) [inline docs/content/preview/api/ysql/the-sql-language/statements/ddl_create_publication.md:68]:
```suggestion
In this example, the publication `puballtables` includes all the current as well as future tables.
```

File: docs/content/preview/api/ysql/the-sql-language/statements/ddl_drop_publication.md [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+
+## Syntax
+
+{{%ebnf%}}
+  drop_publication,
+{{%/ebnf%}}
+
+## Semantics
+
+Drop a publication named *publication_name*. If `publication_name` doesn't already exists in the specified database, an error will be raised unless the `IF EXISTS` clause is used.
```
Comment (ddhodge) [inline docs/content/preview/api/ysql/the-sql-language/statements/ddl_drop_publication.md:27]:
```suggestion
Drop a publication named *publication_name*. If `publication_name` doesn't exist in the specified database, an error will be raised unless the `IF EXISTS` clause is used.
```

File: docs/content/preview/api/ysql/the-sql-language/statements/streaming_drop_repl_slot.md [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+
+## Semantics
+
+### *slot_name*
+
+The name of the replication slot.
+
+## Examples
+
+- Drop a replication slot.
```
Comment (ddhodge) [inline docs/content/preview/api/ysql/the-sql-language/statements/streaming_drop_repl_slot.md:31]:
```suggestion
Drop a replication slot.
```

File: docs/content/preview/api/ysql/the-sql-language/statements/streaming_start_replication.md [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+### *start_replication_option_name*
+
+The name of an option passed to the slot's logical decoding plugin.
+
+### *start_replication_option_value*
+
+Optional value, in the form of a string constant, associated with the specified option.
+
+## Example
+
```
Comment (ddhodge) [inline docs/content/preview/api/ysql/the-sql-language/statements/streaming_start_replication.md:46]: Add brief summary - "Start replication from Replication Slot *test_slot* using LSN *0/2* and something something" (explain what is going on)

**Comment by ddhodge** [comment]
@ddorian can you take a look

File: docs/content/preview/api/ysql/the-sql-language/statements/streaming_start_replication.md (hunk 13 above)
Comment (dr0pdb) [reply docs/content/preview/api/ysql/the-sql-language/statements/streaming_start_replication.md:46]: Added

Comment (ddhodge) [reply docs/content/preview/api/ysql/the-sql-language/statements/streaming_start_replication.md:46]: But what is 'pub'?

Comment (ddorian) [reply docs/content/preview/api/ysql/the-sql-language/statements/streaming_start_replication.md:46]: Do we need to document these options too https://www.postgresql.org/docs/current/protocol-logical-replication.html ?

Comment (ddorian) [reply docs/content/preview/api/ysql/the-sql-language/statements/streaming_start_replication.md:46]: In this example, we probably need a full example? So creating a table, a publication, and then starting replication.
And maybe even show the output of start_replication.

**Review by asrinivasanyb — APPROVED** [review]

File: docs/content/preview/api/ysql/the-sql-language/statements/streaming_start_replication.md (hunk 13 above)
Comment (dr0pdb) [reply docs/content/preview/api/ysql/the-sql-language/statements/streaming_start_replication.md:52]: @ddorian 

> Do we need to document these options too https://www.postgresql.org/docs/current/protocol-logical-replication.html ?

So these parameters depend on the output plugin chosen. These parameters are specific to `pgoutput`. I've added links to the respective documentation for `pgoutput`, `yboutput` and `wal2json`. If the user uses any other plugin, they have to find the documentation for that plugin.

> In this example, we probably need a full example? So creating a table, a publication, and then starting replication.
And maybe even show the output of start_replication.

Good point. I've added one. We can't show the output of the start_replication as it streams data via copy protocol. The command doesn't output anything on the shell.

File: docs/content/preview/api/ysql/the-sql-language/statements/streaming_start_replication.md [hunk 14]
```diff
@@ -38,18 +38,36 @@ The log sequence number from where to start the streaming from.
 
 The name of an option passed to the slot's logical decoding plugin.
 
+The applicable options accepted by the command depends on the output plugin of the replication slot. They can be viewed in the respective documentation of the output plugin itself.
+
+For `pgoutput` and `yboutput`, check the section [53.5.1. Logical Streaming Replication Parameters](https://www.postgresql.org/docs/11/protocol-logical-replication.html) in the PG documentation.
```
Comment (ddorian) [inline docs/content/preview/api/ysql/the-sql-language/statements/streaming_start_replication.md:43]: Do we need to document somehow `yboutput`? Or say how/if it differes from `pgoutput` ?

File: docs/content/preview/api/ysql/the-sql-language/statements/streaming_start_replication.md [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+Create a replication slot with name `test_slot` and output plugin `pgoutput`.
+
+```sql
+yugabyte=# CREATE_REPLICATION_SLOT test_slot LOGICAL pgoutput;
+```
+
+Start replication from the `test_slot` replication slot starting from LSN `0/2`. We also pass the `publication_names` parameter with value `mypublication` to the output plugin. `publication_names` is an output plugin parameter that is accepted by both `pgoutput` and `yboutput` to determine the tables to stream the data from.
+
+```sql
+yugabyte=# START_REPLICATION test_slot LOGICAL 0/2 publication_names 'mypublication';
```
Comment (ddorian) [inline docs/content/preview/api/ysql/the-sql-language/statements/streaming_start_replication.md:70]: Is there any "popular" not-happy path that users might run into when configuring? Like "common errors". Not basic ones like you don't have permission, or table doesn't exist.

**Review by ddorian — APPROVED** [review]
LGTM

File: docs/content/preview/api/ysql/the-sql-language/statements/streaming_start_replication.md (hunk 14 above)
Comment (dr0pdb) [reply docs/content/preview/api/ysql/the-sql-language/statements/streaming_start_replication.md:43]: Yes but not in the syntax section. cc: @asrinivasanyb

File: docs/content/preview/api/ysql/the-sql-language/statements/streaming_start_replication.md (hunk 15 above)
Comment (dr0pdb) [reply docs/content/preview/api/ysql/the-sql-language/statements/streaming_start_replication.md:70]: Not that I am aware of actually. We can add more if we encounter but mostly no.

**Pull request merged**